<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 8 Pooled p-values | SWAG Workshops Repository</title>
  <meta name="description" content="Chapter 8 Pooled p-values | SWAG Workshops Repository" />
  <meta name="generator" content="bookdown 0.31 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 8 Pooled p-values | SWAG Workshops Repository" />
  <meta property="og:type" content="book" />
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 8 Pooled p-values | SWAG Workshops Repository" />
  
  
  

<meta name="author" content="UW Statistical Workshops and Applications Group" />



  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="introduction-to-causal-inference.html"/>
<link rel="next" href="references.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">SWAG Workshops Repository</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a></li>
<li class="chapter" data-level="2" data-path="chapter2.html"><a href="chapter2.html"><i class="fa fa-check"></i><b>2</b> Sampling-Resampling Methods</a>
<ul>
<li class="chapter" data-level="2.1" data-path="chapter2.html"><a href="chapter2.html#introduction-1"><i class="fa fa-check"></i><b>2.1</b> Introduction</a></li>
<li class="chapter" data-level="2.2" data-path="chapter2.html"><a href="chapter2.html#rejection-sampling"><i class="fa fa-check"></i><b>2.2</b> Rejection Sampling</a></li>
<li class="chapter" data-level="2.3" data-path="chapter2.html"><a href="chapter2.html#sampling-resampling-methods"><i class="fa fa-check"></i><b>2.3</b> Sampling-Resampling Methods</a>
<ul>
<li class="chapter" data-level="2.3.1" data-path="chapter2.html"><a href="chapter2.html#overview-of-the-sampling-importance-resampling-algorithm"><i class="fa fa-check"></i><b>2.3.1</b> Overview of the Sampling-Importance-Resampling Algorithm</a></li>
<li class="chapter" data-level="2.3.2" data-path="chapter2.html"><a href="chapter2.html#computational-considerations"><i class="fa fa-check"></i><b>2.3.2</b> Computational Considerations</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="chapter2.html"><a href="chapter2.html#implementation-of-sampling-resampling-methods"><i class="fa fa-check"></i><b>2.4</b> Implementation of Sampling-Resampling Methods</a>
<ul>
<li class="chapter" data-level="2.4.1" data-path="chapter2.html"><a href="chapter2.html#illustrative-example-with-binary-data"><i class="fa fa-check"></i><b>2.4.1</b> Illustrative Example with Binary Data</a></li>
<li class="chapter" data-level="2.4.2" data-path="chapter2.html"><a href="chapter2.html#practical-considerations-for-choosing-proposal-distributions"><i class="fa fa-check"></i><b>2.4.2</b> Practical Considerations for Choosing Proposal Distributions</a></li>
<li class="chapter" data-level="2.4.3" data-path="chapter2.html"><a href="chapter2.html#comparing-proposal-distributions"><i class="fa fa-check"></i><b>2.4.3</b> Comparing Proposal Distributions</a></li>
</ul></li>
<li class="chapter" data-level="2.5" data-path="chapter2.html"><a href="chapter2.html#sampling-resampling-in-multiple-dimensions"><i class="fa fa-check"></i><b>2.5</b> Sampling-Resampling in Multiple Dimensions</a>
<ul>
<li class="chapter" data-level="2.5.1" data-path="chapter2.html"><a href="chapter2.html#exercise-with-illustrative-example"><i class="fa fa-check"></i><b>2.5.1</b> Exercise with Illustrative Example</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="bernstein-von-mises-theorem.html"><a href="bernstein-von-mises-theorem.html"><i class="fa fa-check"></i><b>3</b> Bernstein-von Mises Theorem</a>
<ul>
<li class="chapter" data-level="3.1" data-path="bernstein-von-mises-theorem.html"><a href="bernstein-von-mises-theorem.html#introduction-2"><i class="fa fa-check"></i><b>3.1</b> Introduction</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="bernstein-von-mises-theorem.html"><a href="bernstein-von-mises-theorem.html#bayesian-inference"><i class="fa fa-check"></i><b>3.1.1</b> Bayesian Inference</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="bernstein-von-mises-theorem.html"><a href="bernstein-von-mises-theorem.html#theorem"><i class="fa fa-check"></i><b>3.2</b> Theorem</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="bernstein-von-mises-theorem.html"><a href="bernstein-von-mises-theorem.html#importance"><i class="fa fa-check"></i><b>3.2.1</b> Importance</a></li>
<li class="chapter" data-level="3.2.2" data-path="bernstein-von-mises-theorem.html"><a href="bernstein-von-mises-theorem.html#required-assumptions"><i class="fa fa-check"></i><b>3.2.2</b> Required Assumptions</a></li>
<li class="chapter" data-level="3.2.3" data-path="bernstein-von-mises-theorem.html"><a href="bernstein-von-mises-theorem.html#example-1---normal-normal-model"><i class="fa fa-check"></i><b>3.2.3</b> Example 1 - Normal-normal model</a></li>
<li class="chapter" data-level="3.2.4" data-path="bernstein-von-mises-theorem.html"><a href="bernstein-von-mises-theorem.html#example-2---bernoulli-beta-model"><i class="fa fa-check"></i><b>3.2.4</b> Example 2 - Bernoulli-Beta Model</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="bernstein-von-mises-theorem.html"><a href="bernstein-von-mises-theorem.html#limitations"><i class="fa fa-check"></i><b>3.3</b> Limitations</a>
<ul>
<li class="chapter" data-level="3.3.1" data-path="bernstein-von-mises-theorem.html"><a href="bernstein-von-mises-theorem.html#other-thoughts-on-consistency"><i class="fa fa-check"></i><b>3.3.1</b> Other thoughts on consistency</a></li>
<li class="chapter" data-level="3.3.2" data-path="bernstein-von-mises-theorem.html"><a href="bernstein-von-mises-theorem.html#example-3---prior-has-zero-density-at-theta_0"><i class="fa fa-check"></i><b>3.3.2</b> Example 3 - Prior has zero density at <span class="math inline">\(\theta_0\)</span></a></li>
<li class="chapter" data-level="3.3.3" data-path="bernstein-von-mises-theorem.html"><a href="bernstein-von-mises-theorem.html#example-4---true-parameter-value-is-on-the-boundary"><i class="fa fa-check"></i><b>3.3.3</b> Example 4 - True parameter value is on the boundary</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="variational-inference.html"><a href="variational-inference.html"><i class="fa fa-check"></i><b>4</b> Variational Inference</a>
<ul>
<li class="chapter" data-level="4.1" data-path="variational-inference.html"><a href="variational-inference.html#introduction-3"><i class="fa fa-check"></i><b>4.1</b> Introduction</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="variational-inference.html"><a href="variational-inference.html#frequentist-setting"><i class="fa fa-check"></i><b>4.1.1</b> Frequentist Setting</a></li>
<li class="chapter" data-level="4.1.2" data-path="variational-inference.html"><a href="variational-inference.html#bayesian-setting"><i class="fa fa-check"></i><b>4.1.2</b> Bayesian Setting</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="variational-inference.html"><a href="variational-inference.html#example-mixture-of-gaussians"><i class="fa fa-check"></i><b>4.2</b> Example: Mixture of Gaussians</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="variational-inference.html"><a href="variational-inference.html#coordinate-ascent-mean-field-variational-inference"><i class="fa fa-check"></i><b>4.2.1</b> Coordinate Ascent Mean-Field Variational Inference</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="variational-inference.html"><a href="variational-inference.html#example-stochastic-variational-inference-using-pyro-in-python"><i class="fa fa-check"></i><b>4.3</b> Example: Stochastic Variational Inference using Pyro in Python</a></li>
<li class="chapter" data-level="4.4" data-path="variational-inference.html"><a href="variational-inference.html#takeaways"><i class="fa fa-check"></i><b>4.4</b> Takeaways</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="smoothing-techniques.html"><a href="smoothing-techniques.html"><i class="fa fa-check"></i><b>5</b> Smoothing Techniques</a>
<ul>
<li class="chapter" data-level="5.1" data-path="smoothing-techniques.html"><a href="smoothing-techniques.html#introduction-4"><i class="fa fa-check"></i><b>5.1</b> Introduction</a></li>
<li class="chapter" data-level="5.2" data-path="smoothing-techniques.html"><a href="smoothing-techniques.html#kernel-smoothing-methods"><i class="fa fa-check"></i><b>5.2</b> Kernel Smoothing Methods</a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="smoothing-techniques.html"><a href="smoothing-techniques.html#local-linear-regression"><i class="fa fa-check"></i><b>5.2.1</b> Local linear regression</a></li>
<li class="chapter" data-level="5.2.2" data-path="smoothing-techniques.html"><a href="smoothing-techniques.html#tuning-parameter-bandwidth-selection"><i class="fa fa-check"></i><b>5.2.2</b> Tuning Parameter (bandwidth) Selection</a></li>
<li class="chapter" data-level="5.2.3" data-path="smoothing-techniques.html"><a href="smoothing-techniques.html#extension-and-example-local-logistic-regression"><i class="fa fa-check"></i><b>5.2.3</b> Extension and example: Local logistic regression</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="smoothing-techniques.html"><a href="smoothing-techniques.html#smoothing-spline"><i class="fa fa-check"></i><b>5.3</b> Smoothing Spline</a>
<ul>
<li class="chapter" data-level="5.3.1" data-path="smoothing-techniques.html"><a href="smoothing-techniques.html#computation-1"><i class="fa fa-check"></i><b>5.3.1</b> Computation</a></li>
<li class="chapter" data-level="5.3.2" data-path="smoothing-techniques.html"><a href="smoothing-techniques.html#tuning-parameter-selection"><i class="fa fa-check"></i><b>5.3.2</b> Tuning Parameter Selection</a></li>
<li class="chapter" data-level="5.3.3" data-path="smoothing-techniques.html"><a href="smoothing-techniques.html#extension-and-example-nonparametric-logistic-regression"><i class="fa fa-check"></i><b>5.3.3</b> Extension and example: Nonparametric logistic regression</a></li>
<li class="chapter" data-level="5.3.4" data-path="smoothing-techniques.html"><a href="smoothing-techniques.html#take-home-note"><i class="fa fa-check"></i><b>5.3.4</b> Take-home note</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="analyzing-surveys-with-r.html"><a href="analyzing-surveys-with-r.html"><i class="fa fa-check"></i><b>6</b> Analyzing Surveys with R</a>
<ul>
<li class="chapter" data-level="6.1" data-path="analyzing-surveys-with-r.html"><a href="analyzing-surveys-with-r.html#acknowlegdements"><i class="fa fa-check"></i><b>6.1</b> Acknowlegdements</a></li>
<li class="chapter" data-level="6.2" data-path="analyzing-surveys-with-r.html"><a href="analyzing-surveys-with-r.html#introduction-to-thinking-about-surveys"><i class="fa fa-check"></i><b>6.2</b> Introduction to Thinking About Surveys</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="analyzing-surveys-with-r.html"><a href="analyzing-surveys-with-r.html#total-survey-error-tse"><i class="fa fa-check"></i><b>6.2.1</b> Total Survey Error (TSE)</a></li>
<li class="chapter" data-level="6.2.2" data-path="analyzing-surveys-with-r.html"><a href="analyzing-surveys-with-r.html#probability-vs.-non-probability-surveys"><i class="fa fa-check"></i><b>6.2.2</b> Probability vs. Non-Probability Surveys</a></li>
<li class="chapter" data-level="6.2.3" data-path="analyzing-surveys-with-r.html"><a href="analyzing-surveys-with-r.html#design-based-vs.-model-based-inference"><i class="fa fa-check"></i><b>6.2.3</b> Design-based vs. Model-based Inference</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="analyzing-surveys-with-r.html"><a href="analyzing-surveys-with-r.html#probability-survey-without-non-response-design-based"><i class="fa fa-check"></i><b>6.3</b> Probability Survey without Non-response (design-based)</a>
<ul>
<li class="chapter" data-level="6.3.1" data-path="analyzing-surveys-with-r.html"><a href="analyzing-surveys-with-r.html#some-notation"><i class="fa fa-check"></i><b>6.3.1</b> Some notation</a></li>
<li class="chapter" data-level="6.3.2" data-path="analyzing-surveys-with-r.html"><a href="analyzing-surveys-with-r.html#horvitz-thompson-estimator"><i class="fa fa-check"></i><b>6.3.2</b> Horvitz-Thompson Estimator</a></li>
<li class="chapter" data-level="6.3.3" data-path="analyzing-surveys-with-r.html"><a href="analyzing-surveys-with-r.html#variance-estimation"><i class="fa fa-check"></i><b>6.3.3</b> Variance Estimation</a></li>
<li class="chapter" data-level="6.3.4" data-path="analyzing-surveys-with-r.html"><a href="analyzing-surveys-with-r.html#simple-random-sampling-without-replacement-srswor"><i class="fa fa-check"></i><b>6.3.4</b> Simple Random Sampling Without Replacement (SRSwor)</a></li>
<li class="chapter" data-level="6.3.5" data-path="analyzing-surveys-with-r.html"><a href="analyzing-surveys-with-r.html#stratified-simple-random-sampling-without-replacement-ssrswor"><i class="fa fa-check"></i><b>6.3.5</b> Stratified Simple Random Sampling Without Replacement (SSRSwor)</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="analyzing-surveys-with-r.html"><a href="analyzing-surveys-with-r.html#post-adjustment-model-assisted"><i class="fa fa-check"></i><b>6.4</b> Post-adjustment (model-assisted)</a>
<ul>
<li class="chapter" data-level="6.4.1" data-path="analyzing-surveys-with-r.html"><a href="analyzing-surveys-with-r.html#post-stratification"><i class="fa fa-check"></i><b>6.4.1</b> Post-stratification</a></li>
<li class="chapter" data-level="6.4.2" data-path="analyzing-surveys-with-r.html"><a href="analyzing-surveys-with-r.html#calibration-dealing-with-multiple-auxilliary-variables"><i class="fa fa-check"></i><b>6.4.2</b> Calibration: Dealing with multiple auxilliary variables</a></li>
<li class="chapter" data-level="6.4.3" data-path="analyzing-surveys-with-r.html"><a href="analyzing-surveys-with-r.html#what-makes-a-good-calibration-variable"><i class="fa fa-check"></i><b>6.4.3</b> What makes a good calibration variable?</a></li>
<li class="chapter" data-level="6.4.4" data-path="analyzing-surveys-with-r.html"><a href="analyzing-surveys-with-r.html#variance-estimation-resampling"><i class="fa fa-check"></i><b>6.4.4</b> Variance Estimation: Resampling</a></li>
</ul></li>
<li class="chapter" data-level="6.5" data-path="analyzing-surveys-with-r.html"><a href="analyzing-surveys-with-r.html#non-probability-surveys-model-based"><i class="fa fa-check"></i><b>6.5</b> Non-probability Surveys (model-based)</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="introduction-to-causal-inference.html"><a href="introduction-to-causal-inference.html"><i class="fa fa-check"></i><b>7</b> Introduction to Causal Inference</a>
<ul>
<li class="chapter" data-level="7.1" data-path="introduction-to-causal-inference.html"><a href="introduction-to-causal-inference.html#basic-concepts"><i class="fa fa-check"></i><b>7.1</b> Basic Concepts</a>
<ul>
<li class="chapter" data-level="7.1.1" data-path="introduction-to-causal-inference.html"><a href="introduction-to-causal-inference.html#notation"><i class="fa fa-check"></i><b>7.1.1</b> Notation</a></li>
<li class="chapter" data-level="7.1.2" data-path="introduction-to-causal-inference.html"><a href="introduction-to-causal-inference.html#causal-effect"><i class="fa fa-check"></i><b>7.1.2</b> Causal Effect</a></li>
<li class="chapter" data-level="7.1.3" data-path="introduction-to-causal-inference.html"><a href="introduction-to-causal-inference.html#assumptions"><i class="fa fa-check"></i><b>7.1.3</b> Assumptions</a></li>
<li class="chapter" data-level="7.1.4" data-path="introduction-to-causal-inference.html"><a href="introduction-to-causal-inference.html#propensity-score"><i class="fa fa-check"></i><b>7.1.4</b> Propensity Score</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="introduction-to-causal-inference.html"><a href="introduction-to-causal-inference.html#commonly-used-methods"><i class="fa fa-check"></i><b>7.2</b> Commonly Used Methods</a>
<ul>
<li class="chapter" data-level="7.2.1" data-path="introduction-to-causal-inference.html"><a href="introduction-to-causal-inference.html#matching"><i class="fa fa-check"></i><b>7.2.1</b> Matching</a></li>
<li class="chapter" data-level="7.2.2" data-path="introduction-to-causal-inference.html"><a href="introduction-to-causal-inference.html#stratification"><i class="fa fa-check"></i><b>7.2.2</b> Stratification</a></li>
<li class="chapter" data-level="7.2.3" data-path="introduction-to-causal-inference.html"><a href="introduction-to-causal-inference.html#inverse-probability-weighting"><i class="fa fa-check"></i><b>7.2.3</b> Inverse Probability Weighting</a></li>
<li class="chapter" data-level="7.2.4" data-path="introduction-to-causal-inference.html"><a href="introduction-to-causal-inference.html#doubly-robust-estimation"><i class="fa fa-check"></i><b>7.2.4</b> Doubly Robust Estimation</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="pooled-p-values.html"><a href="pooled-p-values.html"><i class="fa fa-check"></i><b>8</b> Pooled p-values</a>
<ul>
<li class="chapter" data-level="8.1" data-path="pooled-p-values.html"><a href="pooled-p-values.html#a-motivating-example"><i class="fa fa-check"></i><b>8.1</b> A motivating example</a>
<ul>
<li class="chapter" data-level="8.1.1" data-path="pooled-p-values.html"><a href="pooled-p-values.html#some-real-data"><i class="fa fa-check"></i><b>8.1.1</b> Some “real” data</a></li>
</ul></li>
<li class="chapter" data-level="8.2" data-path="pooled-p-values.html"><a href="pooled-p-values.html#pooling-p-values"><i class="fa fa-check"></i><b>8.2</b> Pooling <span class="math inline">\(p\)</span>-values</a></li>
<li class="chapter" data-level="8.3" data-path="pooled-p-values.html"><a href="pooled-p-values.html#adjusting-for-dependence"><i class="fa fa-check"></i><b>8.3</b> Adjusting for dependence</a>
<ul>
<li class="chapter" data-level="8.3.1" data-path="pooled-p-values.html"><a href="pooled-p-values.html#the-wrong-way"><i class="fa fa-check"></i><b>8.3.1</b> The wrong way</a></li>
<li class="chapter" data-level="8.3.2" data-path="pooled-p-values.html"><a href="pooled-p-values.html#a-better-way"><i class="fa fa-check"></i><b>8.3.2</b> A better way</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="9" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i><b>9</b> References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">SWAG Workshops Repository</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="pooled-p-values" class="section level1 hasAnchor" number="8">
<h1><span class="header-section-number">Chapter 8</span> Pooled p-values<a href="pooled-p-values.html#pooled-p-values" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p><em>Chris Salahub</em></p>
<div id="a-motivating-example" class="section level2 hasAnchor" number="8.1">
<h2><span class="header-section-number">8.1</span> A motivating example<a href="pooled-p-values.html#a-motivating-example" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Let’s say we have collected a bunch of genetic data and would like to identify which genes are related to some observed trait. We’ll start by defining some terms to make describing things easier.</p>
<ul>
<li><strong>DNA</strong>: genetic information that is passed between generations and is stored in several big molecules made up of many small molecules called nucleotides</li>
<li><strong>genome</strong>: all the DNA of an organism</li>
<li><strong>gene</strong>: a segment of DNA that gives the blueprint for a particular protein</li>
<li><strong>chromosomes</strong>: one of the several big molecules making up the genome</li>
<li><strong>marker</strong>: a sequence of nucleotides measured at a known position in DNA</li>
<li><strong>sex</strong>: the creation of an organism in a new generation by recombining DNA from two organisms in the previous generation</li>
<li><strong>recombination</strong>: when DNA is changed slightly by segments of DNA literally crossing each other and switching or chromosomes being inherited independently, pictorially: <img src="figs/Homologous-recombination.jpg" style="width:100.0%" alt="Recombination (from https://www.genome.gov/genetics-glossary/homologous-recombination)" /></li>
<li><strong>centiMorgans</strong>: an additive measure of distance on DNA, for two markers which have probability <span class="math inline">\(q\)</span> of being separated during sex the distance in centiMorgans is <span class="math inline">\(d = 50 \ln \frac{1}{1 - 2q}\)</span></li>
</ul>
<p>So, when we talk about genetic data, we mean <strong>markers</strong> at known positions on <strong>chromosomes</strong> which ideally span the whole <strong>genome</strong> of a sample of organisms. We pair this data with a physical measurement - in humans examples include height, eye colour, hair colour, or the presence of a disease - to determine how much physical variation can be explained by genetic variation.</p>
<p>Physical traits that are linked to genetics are either Mendelian or non-Mendelian (you can guess which came first). Mendelian traits have variation explained by one (or very few) gene(s), take eye colour which is mostly down to the gene <em>OCA2</em>. Non-Mendelian traits, like height, have variation which cannot be attributed to one (or a handful of) genetic region(s) alone.</p>
<p>To identify which measured markers are related to a physical trait, a host of significance tests are typically computed to assay the impact of each. While this is not optimal compared to a linear model, it is frequently the only option for genetic studies which collect hundred of thousands of markers for only a few dozen individuals. Difficulty fitting a model practically means we must filter these markers first.</p>
<p>Before even pointing our finger at a subset of genes, however, we might want to save time by asking whether there is anything interesting happening <em>at all</em> in this collection of tests. That is where pooled <span class="math inline">\(p\)</span>-values come in.</p>
<div id="some-real-data" class="section level3 hasAnchor" number="8.1.1">
<h3><span class="header-section-number">8.1.1</span> Some “real” data<a href="pooled-p-values.html#some-real-data" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Our ideal data set would involve humans. Genomics mostly strives to better understand the genetic sources of human diseases and traits to help create better medicine and understand ourselves more fully. Unfortunately, human data is pretty hard to come by ethically, and mice data is a lot easier to find. The code below pulls some that is freely available from <a href="https://www.informatics.jax.org/">Mouse Genome Informatics</a>.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="pooled-p-values.html#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="do">## this script provides a means of downloading mouse genomic data from the</span></span>
<span id="cb1-2"><a href="pooled-p-values.html#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="do">## mouse genome database (MGD), an online repository of data on mouse</span></span>
<span id="cb1-3"><a href="pooled-p-values.html#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="do">## genetics</span></span>
<span id="cb1-4"><a href="pooled-p-values.html#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="do">## all the data is stored in text files at</span></span>
<span id="cb1-5"><a href="pooled-p-values.html#cb1-5" aria-hidden="true" tabindex="-1"></a>mgiUrl <span class="ot">&lt;-</span> <span class="st">&quot;https://www.informatics.jax.org/downloads/reports/&quot;</span></span>
<span id="cb1-6"><a href="pooled-p-values.html#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="do">## we&#39;ll be looking at data from</span></span>
<span id="cb1-7"><a href="pooled-p-values.html#cb1-7" aria-hidden="true" tabindex="-1"></a>mgiPName <span class="ot">&lt;-</span> <span class="st">&quot;MGI_JAX_BSB_Panel.rpt&quot;</span></span>
<span id="cb1-8"><a href="pooled-p-values.html#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="do">## and for reference we&#39;ll also need the marker lists in    </span></span>
<span id="cb1-9"><a href="pooled-p-values.html#cb1-9" aria-hidden="true" tabindex="-1"></a>mgiMNames <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;MRK_List1.rpt&quot;</span>, <span class="st">&quot;MRK_List2.rpt&quot;</span>)</span>
<span id="cb1-10"><a href="pooled-p-values.html#cb1-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-11"><a href="pooled-p-values.html#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="do">## this function reads in MGI text files, which are basically organized</span></span>
<span id="cb1-12"><a href="pooled-p-values.html#cb1-12" aria-hidden="true" tabindex="-1"></a><span class="do">## as a tab-separated data frame with extra information</span></span>
<span id="cb1-13"><a href="pooled-p-values.html#cb1-13" aria-hidden="true" tabindex="-1"></a>readMGIrpt <span class="ot">&lt;-</span> <span class="cf">function</span>(file) {</span>
<span id="cb1-14"><a href="pooled-p-values.html#cb1-14" aria-hidden="true" tabindex="-1"></a>    raw <span class="ot">&lt;-</span> <span class="fu">scan</span>(file, <span class="at">what =</span> <span class="fu">character</span>(), <span class="at">sep =</span> <span class="st">&quot;</span><span class="sc">\n</span><span class="st">&quot;</span>) <span class="co"># most basic R input</span></span>
<span id="cb1-15"><a href="pooled-p-values.html#cb1-15" aria-hidden="true" tabindex="-1"></a>    leg <span class="ot">&lt;-</span> <span class="fu">which</span>(raw <span class="sc">==</span> <span class="st">&quot;Legend:&quot;</span>) <span class="co"># identify the legend rows</span></span>
<span id="cb1-16"><a href="pooled-p-values.html#cb1-16" aria-hidden="true" tabindex="-1"></a>    lenHead <span class="ot">&lt;-</span> leg <span class="sc">+</span> <span class="dv">4</span> <span class="co"># the +4 is from inspection of the legends</span></span>
<span id="cb1-17"><a href="pooled-p-values.html#cb1-17" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">length</span>(leg) <span class="sc">==</span> <span class="dv">0</span>) { <span class="co"># separate legend</span></span>
<span id="cb1-18"><a href="pooled-p-values.html#cb1-18" aria-hidden="true" tabindex="-1"></a>        leg <span class="ot">&lt;-</span> <span class="fu">which</span>(<span class="fu">grepl</span>(<span class="st">&quot;^CHR&quot;</span>, raw))[<span class="dv">1</span>]</span>
<span id="cb1-19"><a href="pooled-p-values.html#cb1-19" aria-hidden="true" tabindex="-1"></a>        lenHead <span class="ot">&lt;-</span> leg</span>
<span id="cb1-20"><a href="pooled-p-values.html#cb1-20" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb1-21"><a href="pooled-p-values.html#cb1-21" aria-hidden="true" tabindex="-1"></a>    desc <span class="ot">&lt;-</span> <span class="fu">paste</span>(raw[<span class="dv">1</span><span class="sc">:</span>lenHead], <span class="at">collapse =</span> <span class="st">&quot;</span><span class="sc">\n</span><span class="st">&quot;</span>) <span class="co"># data description</span></span>
<span id="cb1-22"><a href="pooled-p-values.html#cb1-22" aria-hidden="true" tabindex="-1"></a>    dat <span class="ot">&lt;-</span> raw[(lenHead<span class="sc">+</span><span class="dv">1</span>)<span class="sc">:</span><span class="fu">length</span>(raw)] <span class="co"># actual data</span></span>
<span id="cb1-23"><a href="pooled-p-values.html#cb1-23" aria-hidden="true" tabindex="-1"></a>    <span class="do">## each row in the data is associated with a reference, the study</span></span>
<span id="cb1-24"><a href="pooled-p-values.html#cb1-24" aria-hidden="true" tabindex="-1"></a>    <span class="do">## or paper that characterized or collected the corresponding</span></span>
<span id="cb1-25"><a href="pooled-p-values.html#cb1-25" aria-hidden="true" tabindex="-1"></a>    <span class="do">## observation, we want to separate these</span></span>
<span id="cb1-26"><a href="pooled-p-values.html#cb1-26" aria-hidden="true" tabindex="-1"></a>    refPos <span class="ot">&lt;-</span> <span class="fu">regexec</span>(<span class="st">&quot;</span><span class="sc">\\</span><span class="st">tJ</span><span class="sc">\\</span><span class="st">:[0-9]+(?:, J</span><span class="sc">\\</span><span class="st">:[0-9]+){0,4}&quot;</span>, dat)</span>
<span id="cb1-27"><a href="pooled-p-values.html#cb1-27" aria-hidden="true" tabindex="-1"></a>    refs <span class="ot">&lt;-</span> <span class="fu">sapply</span>(<span class="fu">regmatches</span>(dat, refPos), <span class="co"># extract references</span></span>
<span id="cb1-28"><a href="pooled-p-values.html#cb1-28" aria-hidden="true" tabindex="-1"></a>                   <span class="cf">function</span>(el) {</span>
<span id="cb1-29"><a href="pooled-p-values.html#cb1-29" aria-hidden="true" tabindex="-1"></a>                       <span class="cf">if</span> (<span class="fu">length</span>(el) <span class="sc">==</span> <span class="dv">0</span>) {</span>
<span id="cb1-30"><a href="pooled-p-values.html#cb1-30" aria-hidden="true" tabindex="-1"></a>                           <span class="st">&quot;&quot;</span></span>
<span id="cb1-31"><a href="pooled-p-values.html#cb1-31" aria-hidden="true" tabindex="-1"></a>                       } <span class="cf">else</span> <span class="fu">gsub</span>(<span class="st">&quot;</span><span class="sc">\\</span><span class="st">t&quot;</span>, <span class="st">&quot;&quot;</span>, el)})</span>
<span id="cb1-32"><a href="pooled-p-values.html#cb1-32" aria-hidden="true" tabindex="-1"></a>    data <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">regmatches</span>(dat, refPos, <span class="at">invert =</span> <span class="cn">TRUE</span>)) <span class="co"># remove refs</span></span>
<span id="cb1-33"><a href="pooled-p-values.html#cb1-33" aria-hidden="true" tabindex="-1"></a>    mat <span class="ot">&lt;-</span> <span class="fu">do.call</span>(rbind, <span class="fu">strsplit</span>(data[data <span class="sc">!=</span> <span class="st">&quot;&quot;</span>], <span class="st">&quot;</span><span class="sc">\\</span><span class="st">t&quot;</span>)) <span class="co"># regularize</span></span>
<span id="cb1-34"><a href="pooled-p-values.html#cb1-34" aria-hidden="true" tabindex="-1"></a>    rwnms <span class="ot">&lt;-</span> mat[<span class="dv">1</span>, <span class="sc">-</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>)] <span class="co"># animal numbers/ids</span></span>
<span id="cb1-35"><a href="pooled-p-values.html#cb1-35" aria-hidden="true" tabindex="-1"></a>    colnms <span class="ot">&lt;-</span> mat[<span class="sc">-</span><span class="dv">1</span>, <span class="dv">3</span>] <span class="co"># symbol field</span></span>
<span id="cb1-36"><a href="pooled-p-values.html#cb1-36" aria-hidden="true" tabindex="-1"></a>    colDesc <span class="ot">&lt;-</span> mat[<span class="sc">-</span><span class="dv">1</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>] <span class="co"># symbol details</span></span>
<span id="cb1-37"><a href="pooled-p-values.html#cb1-37" aria-hidden="true" tabindex="-1"></a>    <span class="fu">colnames</span>(colDesc) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;chr&quot;</span>, <span class="st">&quot;mgiid&quot;</span>, <span class="st">&quot;symbol&quot;</span>) <span class="co"># informative names</span></span>
<span id="cb1-38"><a href="pooled-p-values.html#cb1-38" aria-hidden="true" tabindex="-1"></a>    data <span class="ot">&lt;-</span> <span class="fu">t</span>(mat[<span class="sc">-</span><span class="dv">1</span>,<span class="sc">-</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>)]) <span class="co"># observed values</span></span>
<span id="cb1-39"><a href="pooled-p-values.html#cb1-39" aria-hidden="true" tabindex="-1"></a>    <span class="fu">rownames</span>(data) <span class="ot">&lt;-</span> rwnms <span class="co"># final data formatting</span></span>
<span id="cb1-40"><a href="pooled-p-values.html#cb1-40" aria-hidden="true" tabindex="-1"></a>    <span class="fu">colnames</span>(data) <span class="ot">&lt;-</span> colnms</span>
<span id="cb1-41"><a href="pooled-p-values.html#cb1-41" aria-hidden="true" tabindex="-1"></a>    <span class="fu">list</span>(<span class="at">summary =</span> desc, <span class="co"># return everything in a big list</span></span>
<span id="cb1-42"><a href="pooled-p-values.html#cb1-42" aria-hidden="true" tabindex="-1"></a>         <span class="at">markers =</span> <span class="fu">data.frame</span>(colDesc, <span class="at">ref =</span> refs[<span class="sc">-</span><span class="dv">1</span>]),</span>
<span id="cb1-43"><a href="pooled-p-values.html#cb1-43" aria-hidden="true" tabindex="-1"></a>         <span class="at">data =</span> <span class="fu">as.data.frame</span>(data))</span>
<span id="cb1-44"><a href="pooled-p-values.html#cb1-44" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb1-45"><a href="pooled-p-values.html#cb1-45" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-46"><a href="pooled-p-values.html#cb1-46" aria-hidden="true" tabindex="-1"></a><span class="do">## this function looks at the marker reference data and processes it into</span></span>
<span id="cb1-47"><a href="pooled-p-values.html#cb1-47" aria-hidden="true" tabindex="-1"></a><span class="do">## a form easier for future analysis</span></span>
<span id="cb1-48"><a href="pooled-p-values.html#cb1-48" aria-hidden="true" tabindex="-1"></a>readMGIlists <span class="ot">&lt;-</span> <span class="cf">function</span>(<span class="at">fileList =</span> <span class="fu">paste0</span>(mgiUrl, mgiMNames)) {</span>
<span id="cb1-49"><a href="pooled-p-values.html#cb1-49" aria-hidden="true" tabindex="-1"></a>    lists <span class="ot">&lt;-</span> <span class="fu">lapply</span>(fileList, scan, <span class="co"># similar input to above</span></span>
<span id="cb1-50"><a href="pooled-p-values.html#cb1-50" aria-hidden="true" tabindex="-1"></a>                    <span class="at">what =</span> <span class="fu">character</span>(), <span class="at">sep =</span> <span class="st">&quot;</span><span class="sc">\n</span><span class="st">&quot;</span>)</span>
<span id="cb1-51"><a href="pooled-p-values.html#cb1-51" aria-hidden="true" tabindex="-1"></a>    lists <span class="ot">&lt;-</span> <span class="fu">lapply</span>(lists, gsub,  <span class="co"># replace end of line tabs with periods</span></span>
<span id="cb1-52"><a href="pooled-p-values.html#cb1-52" aria-hidden="true" tabindex="-1"></a>                    <span class="at">pattern =</span> <span class="st">&quot;</span><span class="sc">\t</span><span class="st">$&quot;</span>, <span class="at">replacement=</span> <span class="st">&quot;</span><span class="sc">\t</span><span class="st">.&quot;</span>)</span>
<span id="cb1-53"><a href="pooled-p-values.html#cb1-53" aria-hidden="true" tabindex="-1"></a>    splits <span class="ot">&lt;-</span> <span class="fu">lapply</span>(lists, strsplit, <span class="at">split =</span> <span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>) <span class="co"># split by tabs</span></span>
<span id="cb1-54"><a href="pooled-p-values.html#cb1-54" aria-hidden="true" tabindex="-1"></a>    colnms <span class="ot">&lt;-</span> splits[[<span class="dv">1</span>]][[<span class="dv">1</span>]] <span class="co"># column names</span></span>
<span id="cb1-55"><a href="pooled-p-values.html#cb1-55" aria-hidden="true" tabindex="-1"></a>    data <span class="ot">&lt;-</span> <span class="fu">do.call</span>(rbind, <span class="co"># covert into one big matrix</span></span>
<span id="cb1-56"><a href="pooled-p-values.html#cb1-56" aria-hidden="true" tabindex="-1"></a>                    <span class="fu">lapply</span>(splits, <span class="cf">function</span>(splt) <span class="fu">do.call</span>(rbind, splt[<span class="sc">-</span><span class="dv">1</span>])))</span>
<span id="cb1-57"><a href="pooled-p-values.html#cb1-57" aria-hidden="true" tabindex="-1"></a>    <span class="fu">colnames</span>(data) <span class="ot">&lt;-</span> colnms</span>
<span id="cb1-58"><a href="pooled-p-values.html#cb1-58" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as.data.frame</span>(data)</span>
<span id="cb1-59"><a href="pooled-p-values.html#cb1-59" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb1-60"><a href="pooled-p-values.html#cb1-60" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-61"><a href="pooled-p-values.html#cb1-61" aria-hidden="true" tabindex="-1"></a><span class="do">## using indices and a reference table, process centiMorgan positions</span></span>
<span id="cb1-62"><a href="pooled-p-values.html#cb1-62" aria-hidden="true" tabindex="-1"></a>processcMs <span class="ot">&lt;-</span> <span class="cf">function</span>(inds, tab) {</span>
<span id="cb1-63"><a href="pooled-p-values.html#cb1-63" aria-hidden="true" tabindex="-1"></a>    sel <span class="ot">&lt;-</span> tab[inds] <span class="co"># take indices from table</span></span>
<span id="cb1-64"><a href="pooled-p-values.html#cb1-64" aria-hidden="true" tabindex="-1"></a>    sel[<span class="fu">grepl</span>(<span class="st">&quot;syntenic&quot;</span>, sel)] <span class="ot">&lt;-</span> <span class="st">&quot;Inf&quot;</span></span>
<span id="cb1-65"><a href="pooled-p-values.html#cb1-65" aria-hidden="true" tabindex="-1"></a>    <span class="fu">suppressWarnings</span>(<span class="fu">as.numeric</span>(sel)) <span class="co"># warnings by design</span></span>
<span id="cb1-66"><a href="pooled-p-values.html#cb1-66" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb1-67"><a href="pooled-p-values.html#cb1-67" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-68"><a href="pooled-p-values.html#cb1-68" aria-hidden="true" tabindex="-1"></a><span class="do">## this function goes through a processed panel from the above functions</span></span>
<span id="cb1-69"><a href="pooled-p-values.html#cb1-69" aria-hidden="true" tabindex="-1"></a><span class="do">## and drops the measurements that have un</span></span>
<span id="cb1-70"><a href="pooled-p-values.html#cb1-70" aria-hidden="true" tabindex="-1"></a>filterPanelBycM <span class="ot">&lt;-</span> <span class="cf">function</span>(panel, locs) {</span>
<span id="cb1-71"><a href="pooled-p-values.html#cb1-71" aria-hidden="true" tabindex="-1"></a>    locOrd <span class="ot">&lt;-</span> <span class="fu">order</span>(locs) <span class="co"># order position</span></span>
<span id="cb1-72"><a href="pooled-p-values.html#cb1-72" aria-hidden="true" tabindex="-1"></a>    toKeep <span class="ot">&lt;-</span> <span class="fu">is.finite</span>(locs[locOrd]) <span class="co"># drop NAs and Infs</span></span>
<span id="cb1-73"><a href="pooled-p-values.html#cb1-73" aria-hidden="true" tabindex="-1"></a>    outMrk <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(panel<span class="sc">$</span>markers[locOrd[toKeep],],</span>
<span id="cb1-74"><a href="pooled-p-values.html#cb1-74" aria-hidden="true" tabindex="-1"></a>                         <span class="at">cMs =</span> locs[locOrd[toKeep]])</span>
<span id="cb1-75"><a href="pooled-p-values.html#cb1-75" aria-hidden="true" tabindex="-1"></a>    outMrk <span class="ot">&lt;-</span> outMrk[<span class="fu">order</span>(outMrk<span class="sc">$</span>chr),] <span class="co"># order chromosome</span></span>
<span id="cb1-76"><a href="pooled-p-values.html#cb1-76" aria-hidden="true" tabindex="-1"></a>    <span class="fu">list</span>(<span class="at">summary =</span> panel<span class="sc">$</span>summary,</span>
<span id="cb1-77"><a href="pooled-p-values.html#cb1-77" aria-hidden="true" tabindex="-1"></a>         <span class="at">markers =</span> outMrk, <span class="at">data =</span> panel<span class="sc">$</span>data[, outMrk<span class="sc">$</span>symbol])</span>
<span id="cb1-78"><a href="pooled-p-values.html#cb1-78" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb1-79"><a href="pooled-p-values.html#cb1-79" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-80"><a href="pooled-p-values.html#cb1-80" aria-hidden="true" tabindex="-1"></a><span class="do">## pull the marker reference</span></span>
<span id="cb1-81"><a href="pooled-p-values.html#cb1-81" aria-hidden="true" tabindex="-1"></a>mgiMarkers <span class="ot">&lt;-</span> <span class="fu">readMGIlists</span>() <span class="co"># descriptions of all markers</span></span>
<span id="cb1-82"><a href="pooled-p-values.html#cb1-82" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-83"><a href="pooled-p-values.html#cb1-83" aria-hidden="true" tabindex="-1"></a><span class="do">## pull the panel</span></span>
<span id="cb1-84"><a href="pooled-p-values.html#cb1-84" aria-hidden="true" tabindex="-1"></a>mgiPanel <span class="ot">&lt;-</span> <span class="fu">readMGIrpt</span>(<span class="fu">paste0</span>(mgiUrl, mgiPName))</span>
<span id="cb1-85"><a href="pooled-p-values.html#cb1-85" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-86"><a href="pooled-p-values.html#cb1-86" aria-hidden="true" tabindex="-1"></a><span class="do">## match the marker names back to the marker reference</span></span>
<span id="cb1-87"><a href="pooled-p-values.html#cb1-87" aria-hidden="true" tabindex="-1"></a>mgiPanel.mrkr <span class="ot">&lt;-</span> <span class="fu">match</span>(<span class="fu">names</span>(mgiPanel<span class="sc">$</span>data),</span>
<span id="cb1-88"><a href="pooled-p-values.html#cb1-88" aria-hidden="true" tabindex="-1"></a>                        mgiMarkers<span class="sc">$</span><span class="st">`</span><span class="at">Marker Symbol</span><span class="st">`</span>)</span>
<span id="cb1-89"><a href="pooled-p-values.html#cb1-89" aria-hidden="true" tabindex="-1"></a><span class="do">## get cM positions</span></span>
<span id="cb1-90"><a href="pooled-p-values.html#cb1-90" aria-hidden="true" tabindex="-1"></a>mgiPanel.cMs <span class="ot">&lt;-</span> <span class="fu">processcMs</span>(mgiPanel.mrkr, </span>
<span id="cb1-91"><a href="pooled-p-values.html#cb1-91" aria-hidden="true" tabindex="-1"></a>                           <span class="at">tab =</span> mgiMarkers<span class="sc">$</span><span class="st">`</span><span class="at">cM Position</span><span class="st">`</span>)</span>
<span id="cb1-92"><a href="pooled-p-values.html#cb1-92" aria-hidden="true" tabindex="-1"></a><span class="do">## Infs indicate markers localized to a chromosome but without a</span></span>
<span id="cb1-93"><a href="pooled-p-values.html#cb1-93" aria-hidden="true" tabindex="-1"></a><span class="do">## known cM position, while NAs indicate markers missing from the</span></span>
<span id="cb1-94"><a href="pooled-p-values.html#cb1-94" aria-hidden="true" tabindex="-1"></a><span class="do">## reference file or marked as unknown there</span></span>
<span id="cb1-95"><a href="pooled-p-values.html#cb1-95" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-96"><a href="pooled-p-values.html#cb1-96" aria-hidden="true" tabindex="-1"></a><span class="do">## filter the panels by markers with known positions</span></span>
<span id="cb1-97"><a href="pooled-p-values.html#cb1-97" aria-hidden="true" tabindex="-1"></a>mgiFiltered <span class="ot">&lt;-</span> <span class="fu">filterPanelBycM</span>(mgiPanel, mgiPanel.cMs)</span>
<span id="cb1-98"><a href="pooled-p-values.html#cb1-98" aria-hidden="true" tabindex="-1"></a><span class="do">## and convert the data to numeric values</span></span>
<span id="cb1-99"><a href="pooled-p-values.html#cb1-99" aria-hidden="true" tabindex="-1"></a>mgiFiltered<span class="sc">$</span>data <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">lapply</span>(mgiFiltered<span class="sc">$</span>data,</span>
<span id="cb1-100"><a href="pooled-p-values.html#cb1-100" aria-hidden="true" tabindex="-1"></a>                           <span class="cf">function</span>(mrkr) <span class="fu">as.numeric</span>(mrkr<span class="sc">==</span> <span class="st">&quot;b&quot;</span>)))</span>
<span id="cb1-101"><a href="pooled-p-values.html#cb1-101" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-102"><a href="pooled-p-values.html#cb1-102" aria-hidden="true" tabindex="-1"></a><span class="do">## remove intermediate data</span></span>
<span id="cb1-103"><a href="pooled-p-values.html#cb1-103" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(<span class="at">list =</span> <span class="fu">c</span>(<span class="st">&quot;mgiMarkers&quot;</span>, <span class="st">&quot;mgiPanel.cMs&quot;</span>, <span class="st">&quot;mgiPanel.mrkr&quot;</span>,</span>
<span id="cb1-104"><a href="pooled-p-values.html#cb1-104" aria-hidden="true" tabindex="-1"></a>            <span class="st">&quot;mgiPanel&quot;</span>))</span>
<span id="cb1-105"><a href="pooled-p-values.html#cb1-105" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-106"><a href="pooled-p-values.html#cb1-106" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(mgiFiltered)</span></code></pre></div>
<pre><code>##         Length Class      Mode     
## summary    1   -none-     character
## markers    5   data.frame list     
## data    1598   data.frame list</code></pre>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="pooled-p-values.html#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(mgiFiltered<span class="sc">$</span>data)</span></code></pre></div>
<pre><code>## [1]   94 1598</code></pre>
<p>In the end, after reading and cleaning the data, we have 94 mice each measured at 1598 marker locations. We are still in the unfortunate position of lacking physical measurements for any of the 94 mice on this panel, but can still generate pseudo <span class="math inline">\(p\)</span>-values. For Mendelian traits, we can pick a marker and generate a trait based solely on its value, while non-Mendelian traits will require the selection of numerous locations which may not contribute equally.</p>
<p>For now, let’s keep this as simple as possible by ensuring independence. Chromosomes assort independently, so if we take one marker from each chromosome we should have independent markers.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="pooled-p-values.html#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="do">## match in this context will give the index of the first marker on</span></span>
<span id="cb5-2"><a href="pooled-p-values.html#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="do">## each chromosome</span></span>
<span id="cb5-3"><a href="pooled-p-values.html#cb5-3" aria-hidden="true" tabindex="-1"></a>mgiIndep <span class="ot">&lt;-</span> mgiFiltered<span class="sc">$</span>data[, <span class="fu">match</span>(<span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">19</span>, <span class="st">&quot;X&quot;</span>), mgiFiltered<span class="sc">$</span>markers<span class="sc">$</span>chr)]</span>
<span id="cb5-4"><a href="pooled-p-values.html#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="do">## inspect dependence structure</span></span>
<span id="cb5-5"><a href="pooled-p-values.html#cb5-5" aria-hidden="true" tabindex="-1"></a><span class="fu">image</span>(<span class="fu">cor</span>(mgiIndep))</span></code></pre></div>
<p><img src="bookdownproj_files/figure-html/unnamed-chunk-3-1.png" width="672" /></p>
<p>To simulate a Mendelian trait, we’ll simply add noise to the first marker. For non-Mendelian inheritance, let’s just take the average of all markers and add noise.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="pooled-p-values.html#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">2314</span>)</span>
<span id="cb6-2"><a href="pooled-p-values.html#cb6-2" aria-hidden="true" tabindex="-1"></a>traitMendel <span class="ot">&lt;-</span> mgiIndep[, <span class="dv">1</span>] <span class="sc">+</span> <span class="fu">rnorm</span>(<span class="dv">94</span>, <span class="at">sd =</span> <span class="fl">0.3</span>)</span>
<span id="cb6-3"><a href="pooled-p-values.html#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">density</span>(traitMendel))  <span class="co"># what mendelian traits look like</span></span></code></pre></div>
<p><img src="bookdownproj_files/figure-html/unnamed-chunk-4-1.png" width="672" /></p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="pooled-p-values.html#cb7-1" aria-hidden="true" tabindex="-1"></a>traitNonMend <span class="ot">&lt;-</span> <span class="fu">apply</span>(mgiIndep[, <span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>], <span class="dv">1</span>, mean) <span class="sc">+</span> <span class="fu">rnorm</span>(<span class="dv">94</span>, <span class="at">sd =</span> <span class="fl">0.3</span>)</span>
<span id="cb7-2"><a href="pooled-p-values.html#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">density</span>(traitNonMend))  <span class="co"># what non-mendelian traits look like</span></span></code></pre></div>
<p><img src="bookdownproj_files/figure-html/unnamed-chunk-4-2.png" width="672" /></p>
<p>Finally, we can compute <span class="math inline">\(p\)</span>-values for each marker and the traits based on the correlation between them.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="pooled-p-values.html#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="do">## based on the mendelian trait</span></span>
<span id="cb8-2"><a href="pooled-p-values.html#cb8-2" aria-hidden="true" tabindex="-1"></a>(pMendel <span class="ot">&lt;-</span> <span class="fu">apply</span>(mgiIndep, <span class="dv">2</span>, <span class="cf">function</span>(col) <span class="fu">cor.test</span>(traitMendel, col)<span class="sc">$</span>p.value))</span></code></pre></div>
<pre><code>##     D1Mit475     D2Mit312      D3Mit60     D4Mit149     D5Mit331      D6Mit86 
## 8.140813e-28 5.562233e-01 5.701328e-01 7.242746e-01 7.675526e-01 3.105180e-01 
##      D7Mit21       Fcer2a     D9Mit186    D10Mit166      D11Mit1     D12Mit37 
## 5.065030e-01 5.402953e-01 6.714437e-01 4.083773e-01 4.271063e-02 2.917104e-01 
##    D13Mit158    D14Mit179     D15Mit12     D16Mit32        Tfb1m     D18Mit66 
## 5.489663e-01 2.092645e-01 4.023976e-01 1.364540e-01 1.034232e-01 5.719812e-01 
##     D19Mit32      DXMit26 
## 7.726304e-01 8.591501e-01</code></pre>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="pooled-p-values.html#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="do">## based on the non-mendelian trait</span></span>
<span id="cb10-2"><a href="pooled-p-values.html#cb10-2" aria-hidden="true" tabindex="-1"></a>(pnonMendel <span class="ot">&lt;-</span> <span class="fu">apply</span>(mgiIndep, <span class="dv">2</span>, <span class="cf">function</span>(col) <span class="fu">cor.test</span>(traitNonMend,</span>
<span id="cb10-3"><a href="pooled-p-values.html#cb10-3" aria-hidden="true" tabindex="-1"></a>    col)<span class="sc">$</span>p.value))</span></code></pre></div>
<pre><code>##   D1Mit475   D2Mit312    D3Mit60   D4Mit149   D5Mit331    D6Mit86    D7Mit21 
## 0.84662803 0.13598050 0.09371634 0.19260532 0.07979980 0.40937430 0.69903954 
##     Fcer2a   D9Mit186  D10Mit166    D11Mit1   D12Mit37  D13Mit158  D14Mit179 
## 0.04886178 0.14942358 0.14839598 0.04176343 0.15557356 0.42105667 0.57307268 
##   D15Mit12   D16Mit32      Tfb1m   D18Mit66   D19Mit32    DXMit26 
## 0.82419580 0.51590067 0.10878820 0.18622075 0.09006944 0.81802465</code></pre>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="pooled-p-values.html#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="do">## some null cases</span></span>
<span id="cb12-2"><a href="pooled-p-values.html#cb12-2" aria-hidden="true" tabindex="-1"></a>pNull <span class="ot">&lt;-</span> <span class="fu">replicate</span>(<span class="dv">1000</span>, <span class="fu">apply</span>(mgiIndep, <span class="dv">2</span>, <span class="cf">function</span>(col) <span class="fu">cor.test</span>(<span class="fu">runif</span>(<span class="dv">94</span>),</span>
<span id="cb12-3"><a href="pooled-p-values.html#cb12-3" aria-hidden="true" tabindex="-1"></a>    col)<span class="sc">$</span>p.value))</span>
<span id="cb12-4"><a href="pooled-p-values.html#cb12-4" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">t</span>(pNull))</span></code></pre></div>
<pre><code>##        D1Mit475  D2Mit312    D3Mit60   D4Mit149  D5Mit331   D6Mit86   D7Mit21
## [1,] 0.90978137 0.9069532 0.35006885 0.52959032 0.7380976 0.7379150 0.7759258
## [2,] 0.71128190 0.4672418 0.40962358 0.55101467 0.2559132 0.9183818 0.3258728
## [3,] 0.95942070 0.5427160 0.04764853 0.06797646 0.7925067 0.4583663 0.8189367
## [4,] 0.72433189 0.2469022 0.37670242 0.79942407 0.7073759 0.4569672 0.5144509
## [5,] 0.09880001 0.5690093 0.04219049 0.67880376 0.5211135 0.2129442 0.5705933
## [6,] 0.47118092 0.1661241 0.87974774 0.85674033 0.9537631 0.7554583 0.9164587
##         Fcer2a    D9Mit186 D10Mit166   D11Mit1   D12Mit37  D13Mit158  D14Mit179
## [1,] 0.6541013 0.892077786 0.0941390 0.9816285 0.58747481 0.96585574 0.10030964
## [2,] 0.7740573 0.928080700 0.5653199 0.9420153 0.60334296 0.44579138 0.02940274
## [3,] 0.2356859 0.830787629 0.7194775 0.3474443 0.72685240 0.95146071 0.68194085
## [4,] 0.2079808 0.008370045 0.1754181 0.6098431 0.03384509 0.85023276 0.01777558
## [5,] 0.2412231 0.034974937 0.8137640 0.8319588 0.64930667 0.06533281 0.23257463
## [6,] 0.6922345 0.695358501 0.9634527 0.1120413 0.93632232 0.82351369 0.35253150
##       D15Mit12  D16Mit32     Tfb1m   D18Mit66  D19Mit32    DXMit26
## [1,] 0.9908900 0.3714854 0.5952756 0.76844887 0.6071990 0.72313619
## [2,] 0.7466969 0.9263798 0.8664269 0.61143617 0.3156817 0.06558521
## [3,] 0.7392575 0.6634759 0.7311247 0.68270873 0.3402068 0.45887828
## [4,] 0.7151886 0.6484924 0.6415287 0.68611425 0.8145712 0.11268825
## [5,] 0.9612551 0.1987970 0.8847019 0.93284250 0.5656709 0.55879076
## [6,] 0.9654213 0.8425012 0.7445170 0.06411416 0.3992616 0.49837159</code></pre>
</div>
</div>
<div id="pooling-p-values" class="section level2 hasAnchor" number="8.2">
<h2><span class="header-section-number">8.2</span> Pooling <span class="math inline">\(p\)</span>-values<a href="pooled-p-values.html#pooling-p-values" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Whether inheritance is Mendelian, non-Mendelian, or something else, our selection of markers tested against a trait gives a collection of <span class="math inline">\(M=20\)</span> <span class="math inline">\(p\)</span>-values from 20 hypothesis tests, and we’d like to know if further investigation is warranted for any of them. Directly, we want to test the null hypothesis
<span class="math display">\[H_0 = \cap_{i = 1}^M H_{0i}\]</span>
where <span class="math inline">\(H_{0i}\)</span> is the hypothesis of “no association” the <span class="math inline">\(i\)</span>th marker. Another way of thinking about this is by considering the null distributions of our <span class="math inline">\(M\)</span> <span class="math inline">\(p\)</span>-values <span class="math inline">\(p_1, \dots, p_M\)</span> if there is nothing going on. In this case, we’d expect
<span class="math display">\[H_0 = p_1, p_2, \dots, p_M \overset{\mathrm{iid}}{\sim} U\]</span>
where <span class="math inline">\(U = Unif(0,1)\)</span>. The basic idea of pooled <span class="math inline">\(p\)</span>-values is to think of a function that takes <span class="math inline">\(M\)</span> <span class="math inline">\(p\)</span>-values as an argument and returns a <span class="math inline">\(p\)</span>-value that behaves like a univariate <span class="math inline">\(p\)</span>-value for a test of <span class="math inline">\(H_0\)</span>. There are two kinds both based on the null distributions of <span class="math inline">\(\mathbf{p} = p_1, \dots, p_M\)</span>.</p>
<p>The first kind uses <span class="math inline">\(p_{(k)}\)</span>, the <span class="math inline">\(k\)</span>th order statistic of <span class="math inline">\(\mathbf{p}\)</span>. Under <span class="math inline">\(H_0\)</span>, the probability that <span class="math inline">\(p_{(k)}\)</span> is less than or equal to <span class="math inline">\(q\)</span> is the probability that <span class="math inline">\(k\)</span> or more uniform <span class="math inline">\(p\)</span>-values less than <span class="math inline">\(q\)</span>. In a formula, that is
<span class="math display">\[F_{(k)}(q) = P(p_{(k)} \leq q) = \sum_{l = k}^M q^l (1 - q)^{M - l}.\]</span>
Therefore, <span class="math inline">\(F_{(k)}(p_{(k)}) \sim U\)</span> and we have our first pooled <span class="math inline">\(p\)</span>-value <span class="math inline">\(g_{(k)}(\mathbf{p}) = F_{(k)}(p_{(k)})\)</span>. One example of this is from <span class="citation">Tippett (1931)</span>, which uses <span class="math inline">\(p_{(1)}\)</span> to obtain
<span class="math display">\[g_{Tip}(\mathbf{p}) = 1 - (1 - p_{(1)})^M,\]</span>
the exact version of the commonly used approximation <span class="math inline">\(Mp_{(1)}\)</span> known as the Bonferroni correction.</p>
<p>The second kind of pooled <span class="math inline">\(p\)</span>-value makes use of quantile transformations. For a continuous random variable <span class="math inline">\(X\)</span> with cumulative distribution function (CDF) <span class="math inline">\(F\)</span> and quantile function <span class="math inline">\(F^{-1}\)</span>, <span class="math inline">\(F^{-1}(U) = X\)</span>. If <span class="math inline">\(X\)</span> is known to have a CDF <span class="math inline">\(F_M\)</span> under summation, the quantities
<span class="math display">\[g(\mathbf{p}) = 1 - F_M \left ( \sum_{i = 1}^M F^{-1}(1 - p_i) \right )\]</span>
and
<span class="math display">\[g(\mathbf{p}) = F_M \left ( \sum_{i = 1}^M F^{-1}(p_i) \right )\]</span>
will both behave exactly like a univariate <span class="math inline">\(p\)</span>-value. Obvious choices for <span class="math inline">\(X\)</span> are the normal distribution and gamma distribution because both of these distributions are closed under summation. The majority of proposed pooled <span class="math inline">\(p\)</span>-values follow the first of these quantile transformation methods, including <span class="math inline">\(g_{Fis}\)</span> of <span class="citation">Fisher (1932)</span> (using the <span class="math inline">\(\chi^2_2\)</span> distribution) and <span class="math inline">\(g_{Sto}\)</span> of <span class="citation">Stouffer et al. (1949)</span> (using the <span class="math inline">\(N(0,1)\)</span> distribution). Let’s code all of these up:</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="pooled-p-values.html#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="do">## the order statistic function</span></span>
<span id="cb14-2"><a href="pooled-p-values.html#cb14-2" aria-hidden="true" tabindex="-1"></a>gord <span class="ot">&lt;-</span> <span class="cf">function</span>(p, k) {</span>
<span id="cb14-3"><a href="pooled-p-values.html#cb14-3" aria-hidden="true" tabindex="-1"></a>    pk <span class="ot">&lt;-</span> <span class="fu">sort</span>(p, <span class="at">decreasing =</span> <span class="cn">FALSE</span>)[k]  <span class="co"># kth order statistic</span></span>
<span id="cb14-4"><a href="pooled-p-values.html#cb14-4" aria-hidden="true" tabindex="-1"></a>    <span class="dv">1</span> <span class="sc">-</span> <span class="fu">pbinom</span>(k, <span class="fu">length</span>(p), pk)</span>
<span id="cb14-5"><a href="pooled-p-values.html#cb14-5" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb14-6"><a href="pooled-p-values.html#cb14-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-7"><a href="pooled-p-values.html#cb14-7" aria-hidden="true" tabindex="-1"></a><span class="do">## the quantile transformation function</span></span>
<span id="cb14-8"><a href="pooled-p-values.html#cb14-8" aria-hidden="true" tabindex="-1"></a>gquant <span class="ot">&lt;-</span> <span class="cf">function</span>(p, finv, fm) {</span>
<span id="cb14-9"><a href="pooled-p-values.html#cb14-9" aria-hidden="true" tabindex="-1"></a>    <span class="dv">1</span> <span class="sc">-</span> <span class="fu">fm</span>(<span class="fu">sum</span>(<span class="fu">finv</span>(<span class="dv">1</span> <span class="sc">-</span> p)))</span>
<span id="cb14-10"><a href="pooled-p-values.html#cb14-10" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb14-11"><a href="pooled-p-values.html#cb14-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-12"><a href="pooled-p-values.html#cb14-12" aria-hidden="true" tabindex="-1"></a><span class="do">## make particular instances</span></span>
<span id="cb14-13"><a href="pooled-p-values.html#cb14-13" aria-hidden="true" tabindex="-1"></a>gtip <span class="ot">&lt;-</span> <span class="cf">function</span>(p) <span class="fu">gord</span>(p, <span class="dv">1</span>)</span>
<span id="cb14-14"><a href="pooled-p-values.html#cb14-14" aria-hidden="true" tabindex="-1"></a>gfis <span class="ot">&lt;-</span> <span class="cf">function</span>(p) <span class="fu">gquant</span>(p, <span class="at">finv =</span> <span class="cf">function</span>(p) <span class="fu">qchisq</span>(p, <span class="dv">2</span>), <span class="at">fm =</span> <span class="cf">function</span>(s) <span class="fu">pchisq</span>(s,</span>
<span id="cb14-15"><a href="pooled-p-values.html#cb14-15" aria-hidden="true" tabindex="-1"></a>    <span class="dv">2</span> <span class="sc">*</span> <span class="fu">length</span>(p)))</span>
<span id="cb14-16"><a href="pooled-p-values.html#cb14-16" aria-hidden="true" tabindex="-1"></a>gsto <span class="ot">&lt;-</span> <span class="cf">function</span>(p) <span class="fu">gquant</span>(p, <span class="at">finv =</span> qnorm, <span class="at">fm =</span> <span class="cf">function</span>(s) <span class="fu">pnorm</span>(s, <span class="at">sd =</span> <span class="fu">sqrt</span>(<span class="fu">length</span>(p))))</span>
<span id="cb14-17"><a href="pooled-p-values.html#cb14-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-18"><a href="pooled-p-values.html#cb14-18" aria-hidden="true" tabindex="-1"></a><span class="do">## store together in a list</span></span>
<span id="cb14-19"><a href="pooled-p-values.html#cb14-19" aria-hidden="true" tabindex="-1"></a>pools <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">tippett =</span> gtip, <span class="at">fisher =</span> gfis, <span class="at">stouffer =</span> gsto)</span></code></pre></div>
<p>We can now compute the pooled <span class="math inline">\(p\)</span>-values of our earlier Mendelian and non-Mendelian cases:</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="pooled-p-values.html#cb15-1" aria-hidden="true" tabindex="-1"></a>(poolMendel <span class="ot">&lt;-</span> <span class="fu">lapply</span>(pools, <span class="cf">function</span>(f) <span class="fu">f</span>(pMendel)))</span></code></pre></div>
<pre><code>## $tippett
## [1] 0
## 
## $fisher
## [1] 0
## 
## $stouffer
## [1] 0</code></pre>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="pooled-p-values.html#cb17-1" aria-hidden="true" tabindex="-1"></a>(poolNonMend <span class="ot">&lt;-</span> <span class="fu">lapply</span>(pools, <span class="cf">function</span>(f) <span class="fu">f</span>(pnonMendel)))</span></code></pre></div>
<pre><code>## $tippett
## [1] 0.2025844
## 
## $fisher
## [1] 0.01703636
## 
## $stouffer
## [1] 0.005196858</code></pre>
<p>All methods agree for the first set of <span class="math inline">\(p\)</span>-values (not too surprising given that the first <span class="math inline">\(p\)</span>-value is less than <span class="math inline">\(1 \times 10^{-27}\)</span>), but the non-Mendelian case leads to some interesting differences. Let’s look at the two sets of <span class="math inline">\(p\)</span>-values again:</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="pooled-p-values.html#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rbind</span>(<span class="at">Mendel =</span> <span class="fu">round</span>(pMendel, <span class="dv">2</span>), <span class="st">`</span><span class="at">Non-Mendel</span><span class="st">`</span> <span class="ot">=</span> <span class="fu">round</span>(pnonMendel, <span class="dv">2</span>))</span></code></pre></div>
<pre><code>##            D1Mit475 D2Mit312 D3Mit60 D4Mit149 D5Mit331 D6Mit86 D7Mit21 Fcer2a
## Mendel         0.00     0.56    0.57     0.72     0.77    0.31    0.51   0.54
## Non-Mendel     0.85     0.14    0.09     0.19     0.08    0.41    0.70   0.05
##            D9Mit186 D10Mit166 D11Mit1 D12Mit37 D13Mit158 D14Mit179 D15Mit12
## Mendel         0.67      0.41    0.04     0.29      0.55      0.21     0.40
## Non-Mendel     0.15      0.15    0.04     0.16      0.42      0.57     0.82
##            D16Mit32 Tfb1m D18Mit66 D19Mit32 DXMit26
## Mendel         0.14  0.10     0.57     0.77    0.86
## Non-Mendel     0.52  0.11     0.19     0.09    0.82</code></pre>
<p>When only one marker is responsible for the variation in the trait, a single <span class="math inline">\(p\)</span>-value is highly significant. In the case of many responsible markers the <span class="math inline">\(p\)</span>-values are not so obviously significant, with the smallest one barely less than 0.05. Rather, this case has a greater number of small <span class="math inline">\(p\)</span>-values than we would expect by chance if all are truly uniform and independent, a fact more easily seen by viewing a histogram of the two sets of <span class="math inline">\(p\)</span>-values.</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="pooled-p-values.html#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(pMendel, <span class="at">breaks =</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="at">by =</span> <span class="fl">0.2</span>), <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">12</span>))</span>
<span id="cb21-2"><a href="pooled-p-values.html#cb21-2" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(pnonMendel, <span class="at">breaks =</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="at">by =</span> <span class="fl">0.2</span>), <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">12</span>))</span></code></pre></div>
<p><img src="bookdownproj_files/figure-html/figures-side-1.png" width="50%" /><img src="bookdownproj_files/figure-html/figures-side-2.png" width="50%" /></p>
<p>Though the proof is beyond the scope of this workshop, it turns out the ordering of <span class="math inline">\(g_{Tip}\)</span>, <span class="math inline">\(g_{Fis}\)</span>, and <span class="math inline">\(g_{Sto}\)</span> for the non-Mendelian case reflects a deeper concept. Broadly, <em>any</em> pooled <span class="math inline">\(p\)</span>-value is sensitive to either a large number of weakly significant <span class="math inline">\(p\)</span>-values or a small number of strongly significant <span class="math inline">\(p\)</span>-values. There is a fundamental trade-off between these two cases for any pooled <span class="math inline">\(p\)</span>-value, which we can visualize crudely using the mean and minimum values of each simulated null sample in <code>pNull</code>. The following plot shows the pooled <span class="math inline">\(p\)</span>-value for each method by the mean of the sample, meant to represent the evidence spread broadly. Point sizes for each mean value are scaled by the minimum <span class="math inline">\(p\)</span>-value, meant to represent concentrated evidence in one test.</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="pooled-p-values.html#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="do">## compute medians for each sample</span></span>
<span id="cb22-2"><a href="pooled-p-values.html#cb22-2" aria-hidden="true" tabindex="-1"></a>pNullMean <span class="ot">&lt;-</span> <span class="fu">apply</span>(<span class="fu">t</span>(pNull), <span class="dv">1</span>, mean)</span>
<span id="cb22-3"><a href="pooled-p-values.html#cb22-3" aria-hidden="true" tabindex="-1"></a><span class="do">## and the minimum</span></span>
<span id="cb22-4"><a href="pooled-p-values.html#cb22-4" aria-hidden="true" tabindex="-1"></a>pNullMin <span class="ot">&lt;-</span> <span class="fu">apply</span>(<span class="fu">t</span>(pNull), <span class="dv">1</span>, min)</span>
<span id="cb22-5"><a href="pooled-p-values.html#cb22-5" aria-hidden="true" tabindex="-1"></a><span class="do">## as well as each pooled p-value</span></span>
<span id="cb22-6"><a href="pooled-p-values.html#cb22-6" aria-hidden="true" tabindex="-1"></a>pNullPools <span class="ot">&lt;-</span> <span class="fu">apply</span>(<span class="fu">t</span>(pNull), <span class="dv">1</span>, <span class="cf">function</span>(row) <span class="fu">sapply</span>(pools, <span class="cf">function</span>(f) <span class="fu">f</span>(row)))</span>
<span id="cb22-7"><a href="pooled-p-values.html#cb22-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-8"><a href="pooled-p-values.html#cb22-8" aria-hidden="true" tabindex="-1"></a><span class="do">## plot the pooled p-value by mean of sample</span></span>
<span id="cb22-9"><a href="pooled-p-values.html#cb22-9" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(pNullMean, pNullPools[<span class="st">&quot;fisher&quot;</span>, ], <span class="at">pch =</span> <span class="dv">16</span>, <span class="at">ylab =</span> <span class="st">&quot;Pooled p-value&quot;</span>,</span>
<span id="cb22-10"><a href="pooled-p-values.html#cb22-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">cex =</span> pNullMin <span class="sc">*</span> <span class="dv">4</span> <span class="sc">+</span> <span class="fl">0.3</span>, <span class="at">xlab =</span> <span class="st">&quot;Mean of p-value sample&quot;</span>, <span class="at">col =</span> <span class="fu">adjustcolor</span>(<span class="st">&quot;firebrick&quot;</span>,</span>
<span id="cb22-11"><a href="pooled-p-values.html#cb22-11" aria-hidden="true" tabindex="-1"></a>        <span class="fl">0.3</span>))</span>
<span id="cb22-12"><a href="pooled-p-values.html#cb22-12" aria-hidden="true" tabindex="-1"></a><span class="fu">points</span>(pNullMean, pNullPools[<span class="st">&quot;stouffer&quot;</span>, ], <span class="at">pch =</span> <span class="dv">16</span>, <span class="at">cex =</span> pNullMin <span class="sc">*</span></span>
<span id="cb22-13"><a href="pooled-p-values.html#cb22-13" aria-hidden="true" tabindex="-1"></a>    <span class="dv">4</span> <span class="sc">+</span> <span class="fl">0.3</span>, <span class="at">col =</span> <span class="fu">adjustcolor</span>(<span class="st">&quot;steelblue&quot;</span>, <span class="fl">0.3</span>))</span>
<span id="cb22-14"><a href="pooled-p-values.html#cb22-14" aria-hidden="true" tabindex="-1"></a><span class="fu">points</span>(pNullMean, pNullPools[<span class="st">&quot;tippett&quot;</span>, ], <span class="at">pch =</span> <span class="dv">16</span>, <span class="at">cex =</span> pNullMin <span class="sc">*</span> <span class="dv">4</span> <span class="sc">+</span></span>
<span id="cb22-15"><a href="pooled-p-values.html#cb22-15" aria-hidden="true" tabindex="-1"></a>    <span class="fl">0.3</span>, <span class="at">col =</span> <span class="fu">adjustcolor</span>(<span class="st">&quot;seagreen&quot;</span>, <span class="fl">0.3</span>))</span>
<span id="cb22-16"><a href="pooled-p-values.html#cb22-16" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="at">x =</span> <span class="st">&quot;topleft&quot;</span>, <span class="at">col =</span> <span class="fu">adjustcolor</span>(<span class="fu">c</span>(<span class="st">&quot;firebrick&quot;</span>, <span class="st">&quot;steelblue&quot;</span>, <span class="st">&quot;seagreen&quot;</span>),</span>
<span id="cb22-17"><a href="pooled-p-values.html#cb22-17" aria-hidden="true" tabindex="-1"></a>    <span class="fl">0.3</span>), <span class="at">legend =</span> <span class="fu">c</span>(<span class="fu">expression</span>(g[Fis]), <span class="fu">expression</span>(g[Sto]), <span class="fu">expression</span>(g[Tip])),</span>
<span id="cb22-18"><a href="pooled-p-values.html#cb22-18" aria-hidden="true" tabindex="-1"></a>    <span class="at">pch =</span> <span class="dv">16</span>)</span></code></pre></div>
<p><img src="bookdownproj_files/figure-html/unnamed-chunk-10-1.png" width="672" /></p>
<p>Note how <span class="math inline">\(g_{Sto}\)</span> is increasing almost monotonically in the mean, with large and small points spread evenly throughout the line for each mean value. <span class="math inline">\(g_{Fis}\)</span>, as suggested by its position giving a <span class="math inline">\(p\)</span>-value between <span class="math inline">\(g_{Sto}\)</span> and <span class="math inline">\(g_{Tip}\)</span>, follows this line somewhat, but is somewhat sensitive to the minimum. The large points, indicating samples with large minimum values, tend to have a larger <span class="math inline">\(p\)</span>-value than the <span class="math inline">\(g_{Sto}\)</span> line while the small points, samples with small minimum values, tend to have a smaller <span class="math inline">\(p\)</span>-value. <span class="math inline">\(g_{Tip}\)</span> barely follows the line at all, the only relationship we can see in the green points is between the size of the point and its <span class="math inline">\(y\)</span>-value.</p>
<p>All of this to say:</p>
<ul>
<li><span class="math inline">\(g_{Sto}\)</span> is not sensitive to the minimum, it follows the overall evidence spread among the <span class="math inline">\(p\)</span>-values</li>
<li><span class="math inline">\(g_{Fis}\)</span> balances sensitivity to the minimum and overall spread of <span class="math inline">\(p\)</span>-values</li>
<li><span class="math inline">\(g_{Tip}\)</span> is not sensitive to the overall spread of evidence, it only responds to the minimum</li>
</ul>
<p>Of course, these examples all assume independence of the <span class="math inline">\(p\)</span>-values. In the case of dependence, the idea of what “spread” evidence means is a lot less clear. Are multiple small <span class="math inline">\(p\)</span>-values an indication of multiple independent contributing variables, or simply an artifact of correlations between them?</p>
</div>
<div id="adjusting-for-dependence" class="section level2 hasAnchor" number="8.3">
<h2><span class="header-section-number">8.3</span> Adjusting for dependence<a href="pooled-p-values.html#adjusting-for-dependence" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>A comprehensive overview of the methods used to adjust for dependence can be found in the companion paper to the <code>poolr</code> package, <span class="citation">Cinar and Viechtbauer (2022)</span>. It should come as no surprise that the above functions have been implemented in this package alongside the adjustments. So, instead of using the functions we wrote above, we can write</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="pooled-p-values.html#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(poolr)</span>
<span id="cb23-2"><a href="pooled-p-values.html#cb23-2" aria-hidden="true" tabindex="-1"></a><span class="do">## the naming is more convenient than ours</span></span>
<span id="cb23-3"><a href="pooled-p-values.html#cb23-3" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&quot;Mendelian p-values</span><span class="sc">\n\n</span><span class="st">&quot;</span>)</span></code></pre></div>
<pre><code>## Mendelian p-values</code></pre>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="pooled-p-values.html#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">fisher</span>(pMendel, <span class="at">adjust =</span> <span class="st">&quot;none&quot;</span>)</span></code></pre></div>
<pre><code>## combined p-values with:      Fisher&#39;s method
## number of p-values combined: 20 
## test statistic:              161.118 ~ chi-square(df = 40) 
## adjustment:                  none 
## combined p-value:            1.817574e-16</code></pre>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="pooled-p-values.html#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&quot;-------</span><span class="sc">\n</span><span class="st">&quot;</span>)</span></code></pre></div>
<pre><code>## -------</code></pre>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb29-1"><a href="pooled-p-values.html#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="fu">stouffer</span>(pMendel, <span class="at">adjust =</span> <span class="st">&quot;none&quot;</span>)</span></code></pre></div>
<pre><code>## combined p-values with:      Stouffer&#39;s method
## number of p-values combined: 20 
## test statistic:              2.894 ~ N(0,1) 
## adjustment:                  none 
## combined p-value:            0.00190034</code></pre>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb31-1"><a href="pooled-p-values.html#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&quot;-------</span><span class="sc">\n</span><span class="st">&quot;</span>)</span></code></pre></div>
<pre><code>## -------</code></pre>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb33-1"><a href="pooled-p-values.html#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="fu">tippett</span>(pMendel, <span class="at">adjust =</span> <span class="st">&quot;none&quot;</span>)  <span class="co"># more accurate for small p-values</span></span></code></pre></div>
<pre><code>## combined p-values with:      Tippett&#39;s method
## number of p-values combined: 20 
## minimum p-value:             0 
## adjustment:                  none 
## combined p-value:            0</code></pre>
<p>We might immediately notice that this is more accurate for the small <span class="math inline">\(p\)</span>-value occurring in the Mendelian example. This is a direct result of the way <code>gquant</code> is defined, we can achieve the same result as <code>poolr</code> by using the <code>lower.tail</code> argument rather than subtracting values from 1. Computational statistics is not quite the same as mathematical statistics!</p>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb35-1"><a href="pooled-p-values.html#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="fu">pnorm</span>(<span class="fu">sum</span>(<span class="fu">qnorm</span>(pMendel, <span class="at">lower.tail =</span> <span class="cn">FALSE</span>)), <span class="at">sd =</span> <span class="fu">sqrt</span>(<span class="dv">20</span>), <span class="at">lower.tail =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<pre><code>## [1] 0.00190034</code></pre>
<div id="the-wrong-way" class="section level3 hasAnchor" number="8.3.1">
<h3><span class="header-section-number">8.3.1</span> The wrong way<a href="pooled-p-values.html#the-wrong-way" class="anchor-section" aria-label="Anchor link to header"></a></h3>
</div>
<div id="a-better-way" class="section level3 hasAnchor" number="8.3.2">
<h3><span class="header-section-number">8.3.2</span> A better way<a href="pooled-p-values.html#a-better-way" class="anchor-section" aria-label="Anchor link to header"></a></h3>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="introduction-to-causal-inference.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="references.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/uwswagclub/workshop-bookdown/edit/master/08-pooled-p-values.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": "https://github.com/uwswagclub/workshop-bookdown/blob/master/08-pooled-p-values.Rmd",
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
